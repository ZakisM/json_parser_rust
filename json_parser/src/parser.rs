use bumpalo::{collections::Vec, Bump};

use crate::{
    ast::{JsonProperty, JsonValue},
    error::ExpectedTokenError,
    token::{Lexer, Token, TokenKind},
};

macro_rules! expected_token_err {
    ($actual_token:expr, $row:expr, $column:expr, $expected_token:path) => {
        return Err(ExpectedTokenError::with_offset(
            vec![$expected_token],
            $actual_token.kind,
            ($actual_token.origin).to_owned(),
            $row,
            $column,
        ))
    };
    ($actual_token:expr, $row:expr, $column:expr, $( $variant:ident )|+) => {
        return Err(ExpectedTokenError::with_offset(
            vec![$(TokenKind::$variant),+],
            $actual_token.kind,
            ($actual_token.origin).to_owned(),
            $row,
            $column,
        ))
    };
}

#[derive(Debug)]
pub struct Parser<'a> {
    lexer: Lexer<'a>,
    current_token: Token<'a>,
    peek_token: Token<'a>,
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        let mut parser = Self {
            lexer: Lexer::new(input),
            current_token: Token::default(),
            peek_token: Token::default(),
        };

        parser.next_token();

        parser
    }

    fn next_token(&mut self) {
        self.current_token = self.peek_token;
        self.peek_token = self.lexer.next_token();
    }

    fn expect_peek(&mut self, expected: TokenKind) -> Result<(), ExpectedTokenError> {
        if self.peek_token.kind != expected {
            expected_token_err!(self.peek_token, self.lexer.row, self.lexer.column, expected)
        }

        self.next_token();

        Ok(())
    }

    fn parse_string(&self, literal: &'a str) -> Result<JsonValue<'a>, ExpectedTokenError> {
        Ok(JsonValue::String(literal))
    }

    fn parse_number(&self, literal: &'a str) -> Result<JsonValue<'a>, ExpectedTokenError> {
        let n = literal.parse::<f64>().map_err(|_| {
            ExpectedTokenError::with_offset(
                vec![TokenKind::Number],
                TokenKind::Illegal,
                literal.to_owned(),
                self.lexer.row,
                self.lexer.column,
            )
        })?;

        Ok(JsonValue::Number(n))
    }

    fn parse_value(&mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        let value = match &self.peek_token.kind {
            TokenKind::String => self.parse_string(self.peek_token.origin)?,
            TokenKind::Number => self.parse_number(self.peek_token.origin)?,
            TokenKind::True => JsonValue::Boolean(true),
            TokenKind::False => JsonValue::Boolean(false),
            TokenKind::Null => JsonValue::Null,
            TokenKind::LBrace => self.parse_object(bump)?,
            TokenKind::LBracket => self.parse_array(bump)?,
            _ => {
                expected_token_err!(
                    self.peek_token,
                    self.lexer.row,
                    self.lexer.column,
                    String | Number | True | False | Null | LBrace | LBracket
                )
            }
        };
        self.next_token();

        Ok(value)
    }

    fn parse_property(&mut self, bump: &'a Bump) -> Result<JsonProperty<'a>, ExpectedTokenError> {
        self.expect_peek(TokenKind::String)?;

        let key = self.current_token.origin;

        self.expect_peek(TokenKind::Colon)?;

        let value = self.parse_value(bump)?;

        Ok(JsonProperty::from((key, value)))
    }

    fn parse_array(&mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        self.expect_peek(TokenKind::LBracket)?;

        if self.peek_token.kind == TokenKind::RBracket {
            return Ok(JsonValue::Array(Vec::new_in(bump)));
        }

        let mut items = Vec::with_capacity_in(8, bump);

        loop {
            let value = self.parse_value(bump)?;
            items.push(value);

            match &self.peek_token.kind {
                TokenKind::Comma => self.next_token(),
                TokenKind::RBracket => break,
                _ => {
                    expected_token_err!(
                        self.peek_token,
                        self.lexer.row,
                        self.lexer.column,
                        Comma | RBracket
                    )
                }
            }
        }

        Ok(JsonValue::Array(items))
    }

    fn parse_object(&mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        self.expect_peek(TokenKind::LBrace)?;

        if self.peek_token.kind == TokenKind::RBrace {
            return Ok(JsonValue::Object(Vec::new_in(bump)));
        }

        let mut items = Vec::with_capacity_in(8, bump);

        loop {
            let item = self.parse_property(bump)?;
            items.push(item);

            match &self.peek_token.kind {
                TokenKind::Comma => self.next_token(),
                TokenKind::RBrace => break,
                _ => {
                    expected_token_err!(
                        self.peek_token,
                        self.lexer.row,
                        self.lexer.column,
                        Comma | RBrace
                    )
                }
            }
        }

        Ok(JsonValue::Object(items))
    }

    fn parse_root_object(mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        let result = self.parse_object(bump)?;

        self.next_token();

        if !matches!(
            (&self.current_token.kind, &self.peek_token.kind),
            (TokenKind::RBrace, TokenKind::Eof)
        ) {
            expected_token_err!(
                self.peek_token,
                self.lexer.row,
                self.lexer.column,
                TokenKind::Eof
            )
        }

        Ok(result)
    }

    fn parse_root_array(mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        let result = self.parse_array(bump)?;

        self.next_token();

        if !matches!(
            (&self.current_token.kind, &self.peek_token.kind),
            (TokenKind::RBracket, TokenKind::Eof)
        ) {
            expected_token_err!(
                self.peek_token,
                self.lexer.row,
                self.lexer.column,
                TokenKind::Eof
            )
        }

        Ok(result)
    }

    pub fn parse(self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        match self.peek_token.kind {
            TokenKind::LBrace => self.parse_root_object(bump),
            TokenKind::LBracket => self.parse_root_array(bump),
            _ => expected_token_err!(
                self.current_token,
                self.lexer.row,
                self.lexer.column,
                LBrace | LBracket
            ),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use bumpalo::vec as bump_vec;

    #[test]
    fn parse_top_level_object() {
        let json = r#"
{
	"string": "Hello, world!",
	"number": -42,
	"nested_object": {
		"nested_string": "This is a nested string",
		"nested_number": [100, 3.21865081787e-6, 300, [400, -500, [600, [700, {"secret": 12345}]]]]
	},
	"boolean": true,
	"nested_deep_empty_array": [[[]], {}],
	"nested_empties": {
	    "empty_object": {},
	    "empty_array": []
	}
}
"#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Ok(JsonValue::Object(bump_vec![in &bump;
                JsonProperty::from(("string", JsonValue::String("Hello, world!"))),
                JsonProperty::from(("number", JsonValue::Number(-42.0))),
                JsonProperty::from((
                    "nested_object",
                    JsonValue::Object(bump_vec![
                        in &bump;
                        JsonProperty::from((
                            "nested_string",
                            JsonValue::String("This is a nested string")
                        )),
                        JsonProperty::from((
                            "nested_number",
                            JsonValue::Array(bump_vec![
                                in &bump;
                                JsonValue::Number(100.0),
                                JsonValue::Number(3.21865081787e-6),
                                JsonValue::Number(300.0),
                                JsonValue::Array(bump_vec![
                                    in &bump;
                                    JsonValue::Number(400.0),
                                    JsonValue::Number(-500.0),
                                    JsonValue::Array(bump_vec![
                                        in &bump;
                                        JsonValue::Number(600.0),
                                        JsonValue::Array(bump_vec![
                                            in &bump;
                                            JsonValue::Number(700.0),
                                            JsonValue::Object(bump_vec![
                                                in &bump;
                                                JsonProperty::from(("secret", JsonValue::Number(12345.0)
                                            ))])
                                        ])
                                    ])
                                ])
                            ])
                        ))
                    ])
                )),
                JsonProperty::from(("boolean", JsonValue::Boolean(true))),
                JsonProperty::from(("nested_deep_empty_array", JsonValue::Array(bump_vec![
                    in &bump;
                    JsonValue::Array(bump_vec![
                        in &bump;
                        JsonValue::Array(bump_vec![
                            in &bump;
                        ])
                    ]),
                    JsonValue::Object(bump_vec![
                        in &bump;
                    ])
                ]))),
                JsonProperty::from(("nested_empties", JsonValue::Object(bump_vec![
                    in &bump;
                    JsonProperty::from(("empty_object", JsonValue::Object(bump_vec![
                        in &bump;
                    ]))),
                    JsonProperty::from(("empty_array", JsonValue::Array(bump_vec![
                        in &bump;
                    ])))
                ]))),
            ]))
        );
    }

    #[test]
    fn parse_top_level_array() {
        let json = r#"
[
    {
        "one": 1,
        "two": 2
    }
]
"#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Ok(JsonValue::Array(bump_vec![
                in &bump;
                JsonValue::Object(bump_vec![
                    in &bump;
                    JsonProperty::from(("one", JsonValue::Number(1.0))),
                    JsonProperty::from(("two", JsonValue::Number(2.0)))
            ])]))
        );
    }

    #[test]
    fn parse_invalid_number() {
        let json = r#"
    {
        "one": 4eee,
        "two": 2
    }
    "#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Err(ExpectedTokenError {
                expected: vec![TokenKind::Number],
                actual: TokenKind::Illegal,
                origin: "4eee".to_owned(),
                row: 3,
                column: 16
            })
        );
    }

    #[test]
    fn parse_invalid_utf8() {
        let json = r#"
    {
        "one": 4èee,
        "two": 2
    }
    "#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Err(ExpectedTokenError {
                expected: vec![TokenKind::Comma, TokenKind::RBrace],
                actual: TokenKind::Illegal,
                origin: "è".to_owned(),
                row: 3,
                column: 17
            })
        );
    }

    #[test]
    fn parse_invalid_octal() {
        let json = r#"
[-65619720000000.61972000000,029]
    "#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Err(ExpectedTokenError {
                expected: vec![
                    TokenKind::String,
                    TokenKind::Number,
                    TokenKind::True,
                    TokenKind::False,
                    TokenKind::Null,
                    TokenKind::LBrace,
                    TokenKind::LBracket
                ],
                actual: TokenKind::Illegal,
                origin: "0".to_owned(),
                row: 2,
                column: 30
            })
        );
    }
}
