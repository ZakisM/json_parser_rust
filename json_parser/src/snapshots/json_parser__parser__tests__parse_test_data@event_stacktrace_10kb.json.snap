---
source: json_parser/src/parser.rs
expression: parser.parse(&bump)
input_file: test_data/event_stacktrace_10kb.json
---
Ok(
    Object(
        [
            JsonProperty {
                key: "@t",
                value: String(
                    "2020-02-07T19:34:57.3431468Z",
                ),
            },
            JsonProperty {
                key: "@mt",
                value: String(
                    "Connection id “{ConnectionId}”, Request id “{TraceIdentifier}”: An unhandled exception was thrown by the application.",
                ),
            },
            JsonProperty {
                key: "@m",
                value: String(
                    "Connection id \\\"0HLTC0J80T1UV\\\", Request id \\\"0HLTC0J80T1UV:00000001\\\": An unhandled exception was thrown by the application.",
                ),
            },
            JsonProperty {
                key: "@i",
                value: String(
                    "42956a2f",
                ),
            },
            JsonProperty {
                key: "@l",
                value: String(
                    "Error",
                ),
            },
            JsonProperty {
                key: "@x",
                value: String(
                    "System.Security.Cryptography.CryptographicException: An error occurred while trying to encrypt the provided data. Refer to the inner exception for more information.\\n ---> Npgsql.NpgsqlException (0x80004005): Exception while connecting\\n ---> System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (111): Connection refused /sql/myapp-production:myapp/.s.PGSQL.5432\\n   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)\\n   at System.Net.Sockets.Socket.Connect(EndPoint remoteEP)\\n   at Npgsql.NpgsqlConnector.Connect(NpgsqlTimeout timeout)\\n   at Npgsql.NpgsqlConnector.Connect(NpgsqlTimeout timeout)\\n   at Npgsql.NpgsqlConnector.RawOpen(NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)\\n   at Npgsql.NpgsqlConnector.Open(NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)\\n   at Npgsql.ConnectorPool.AllocateLong(NpgsqlConnection conn, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)\\n   at Npgsql.NpgsqlConnection.<>c__DisplayClass32_0.<<Open>g__OpenLong|0>d.MoveNext()\\n--- End of stack trace from previous location where exception was thrown ---\\n   at Npgsql.NpgsqlConnection.Open()\\n   at MyApp.Web.Infrastructure.DataProtection.PostgresDataProtectionXmlRepository.GetAllElements() in /src/MyApp.Web/Infrastructure/DataProtection/PostgresDataProtectionXmlRepository.cs:line 29\\n   at Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager.GetAllKeys()\\n   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded)\\n   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.Microsoft.AspNetCore.DataProtection.KeyManagement.Internal.ICacheableKeyRingProvider.GetCacheableKeyRing(DateTimeOffset now)\\n   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRingCore(DateTime utcNow, Boolean forceRefresh)\\n   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRing()\\n   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.Protect(Byte[] plaintext)\\n   --- End of inner exception stack trace ---\\n   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.Protect(Byte[] plaintext)\\n   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgeryTokenSerializer.Serialize(AntiforgeryToken token)\\n   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgery.Serialize(IAntiforgeryFeature antiforgeryFeature)\\n   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgery.GetAndStoreTokens(HttpContext httpContext)\\n   at Microsoft.AspNetCore.Mvc.ViewFeatures.AntiforgeryExtensions.GetHtml(IAntiforgery antiforgery, HttpContext httpContext)\\n   at Microsoft.AspNetCore.Mvc.ViewFeatures.DefaultHtmlGenerator.GenerateAntiforgery(ViewContext viewContext)\\n   at Microsoft.AspNetCore.Mvc.TagHelpers.FormTagHelper.Process(TagHelperContext context, TagHelperOutput output)\\n   at Microsoft.AspNetCore.Razor.TagHelpers.TagHelper.ProcessAsync(TagHelperContext context, TagHelperOutput output)\\n   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.RunAsync(TagHelperExecutionContext executionContext)\\n   at AspNetCore.Areas_None_Shared__Footer.ExecuteAsync() in /src/MyApp.Web/Areas/None/Shared/_Footer.cshtml:line 30\\n   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)\\n   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)\\n   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)\\n   at Microsoft.AspNetCore.Mvc.TagHelpers.PartialTagHelper.RenderPartialViewAsync(TextWriter writer, Object model, IView view)\\n   at Microsoft.AspNetCore.Mvc.TagHelpers.PartialTagHelper.ProcessAsync(TagHelperContext context, TagHelperOutput output)\\n   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.<RunAsync>g__Awaited|0_0(Task task, TagHelperExecutionContext executionContext, Int32 i, Int32 count)\\n   at AspNetCore.Areas_None_Shared__Layout.<ExecuteAsync>b__19_1() in /src/MyApp.Web/Areas/None/Shared/_Layout.cshtml:line 30\\n   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperExecutionContext.SetOutputContentAsync()\\n   at AspNetCore.Areas_None_Shared__Layout.ExecuteAsync() in /src/MyApp.Web/Areas/None/Shared/_Layout.cshtml:line 24\\n   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)\\n   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)\\n   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderLayoutAsync(ViewContext context, ViewBufferTextWriter bodyWriter)\\n   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)\\n   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)\\n   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)\\n   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ActionContext actionContext, IView view, ViewDataDictionary viewData, ITempDataDictionary tempData, String contentType, Nullable`1 statusCode)\\n   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.ExecuteAsync(ActionContext context, ViewResult result)\\n   at Microsoft.AspNetCore.Mvc.ViewResult.ExecuteResultAsync(ActionContext context)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|29_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()\\n--- End of stack trace from previous location where exception was thrown ---\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()\\n--- End of stack trace from previous location where exception was thrown ---\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)\\n   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)\\n   at MyApp.Web.Areas.None.Shared.Components.Toaster.ToasterMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) in /src/MyApp.Web/Areas/None/Shared/Components/Toaster/ToasterMiddleware.cs:line 43\\n   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()\\n--- End of stack trace from previous location where exception was thrown ---\\n   at MyApp.Web.Infrastructure.Users.CurrentUserMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) in /src/MyApp.Web/Infrastructure/Users/CurrentUserMiddleware.cs:line 23\\n   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()\\n--- End of stack trace from previous location where exception was thrown ---\\n   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)\\n   at Serilog.AspNetCore.RequestLoggingMiddleware.Invoke(HttpContext httpContext)\\n   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)\\n   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()\\n--- End of stack trace from previous location where exception was thrown ---\\n   at MyApp.Web.MyAppWebHost.<>c.<<Build>b__1_14>d.MoveNext() in /src/MyApp.Web/MyAppWebHost.cs:line 155\\n--- End of stack trace from previous location where exception was thrown ---\\n   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)\\n   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.HandleException(HttpContext context, ExceptionDispatchInfo edi)\\n   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__\\/Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)\\n   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)",
                ),
            },
            JsonProperty {
                key: "ConnectionId",
                value: String(
                    "0HLTC0J80T1UV",
                ),
            },
            JsonProperty {
                key: "TraceIdentifier",
                value: String(
                    "0HLTC0J80T1UV:00000001",
                ),
            },
            JsonProperty {
                key: "EventId",
                value: Object(
                    [
                        JsonProperty {
                            key: "Id",
                            value: Number(
                                13.0,
                            ),
                        },
                        JsonProperty {
                            key: "Name",
                            value: String(
                                "ApplicationError",
                            ),
                        },
                    ],
                ),
            },
            JsonProperty {
                key: "SourceContext",
                value: String(
                    "Microsoft.AspNetCore.Server.Kestrel",
                ),
            },
            JsonProperty {
                key: "RequestId",
                value: String(
                    "0HLTC0J80T1UV:00000001",
                ),
            },
            JsonProperty {
                key: "RequestPath",
                value: String(
                    "😎",
                ),
            },
            JsonProperty {
                key: "SpanId",
                value: String(
                    "|16fd0678-4b69eb5e5897888c.",
                ),
            },
            JsonProperty {
                key: "TraceId",
                value: String(
                    "16fd0678-4b69eb5e5897888c",
                ),
            },
            JsonProperty {
                key: "ParentId",
                value: String(
                    "",
                ),
            },
            JsonProperty {
                key: "Application",
                value: String(
                    "MyApp.Web",
                ),
            },
            JsonProperty {
                key: "ImageVersion",
                value: String(
                    "1.0.673",
                ),
            },
            JsonProperty {
                key: "Site",
                value: String(
                    "Production",
                ),
            },
        ],
    ),
)
